File: src/content/docs/topics/cs/coding-problems/advanced-graphs/787-cheapest-flights-within-k-stops.py
def find_cheapest_price(n, flights, src, dst, k):
    pass  # TODO: implement

def _run_tests():
    flights = [[0, 1, 100], [1, 2, 100], [2, 0, 100], [1, 3, 600], [2, 3, 200]]
    assert find_cheapest_price(4, flights, 0, 3, 1) == 700
    assert find_cheapest_price(4, flights, 0, 3, 0) == -1
    assert find_cheapest_price(4, flights, 0, 3, 2) == 400
    assert find_cheapest_price(2, [[0, 1, 500]], 0, 1, 0) == 500
    assert find_cheapest_price(3, [[0, 1, 100], [1, 2, 50]], 1, 1, 1) == 0
    assert find_cheapest_price(3, [[0, 1, 100]], 0, 2, 5) == -1
    # --- large-input timing ---
    import time as _t
    _t0 = _t.perf_counter()
    big_flights = [[i, i + 1, i + 1] for i in range(199)]
    find_cheapest_price(200, big_flights, 0, 199, 200)
    _ms = (_t.perf_counter() - _t0) * 1000
    print(f'perf cheapest-flights 200 nodes chain k=200: {_ms:.1f}ms')
    print('all tests pass')

if __name__ == '__main__':
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/advanced-graphs/787-cheapest-flights-within-k-stops.swift
// LEETCODE_TYPE: Solution
func expectEqual<T: Equatable>(
    _ actual: T,
    _ expected: T,
    _ message: String = "",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard actual == expected else {
        let detail = message.isEmpty ? "values differ" : message
        fatalError("\(file):\(line): \(detail). Expected \(expected), got \(actual)")
    }
}

func expectTrue(
    _ condition: @autoclosure () -> Bool,
    _ message: String = "expected true",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard condition() else {
        fatalError("\(file):\(line): \(message)")
    }
}

func reportSuccess() {
    print("All Swift tests passed")
}

// SWIFT_CATALOG_HELPER: BinaryHeap
struct BinaryHeap<Element> {
    private var elements: [Element] = []
    private let hasHigherPriority: (Element, Element) -> Bool

    init(hasHigherPriority: @escaping (Element, Element) -> Bool) {
        self.hasHigherPriority = hasHigherPriority
    }

    var count: Int { elements.count }
    var isEmpty: Bool { elements.isEmpty }
    var peek: Element? { elements.first }

    mutating func insert(_ element: Element) {
        elements.append(element)
        siftUp(from: elements.count - 1)
    }

    mutating func removeRoot() -> Element? {
        guard !elements.isEmpty else { return nil }
        if elements.count == 1 { return elements.removeLast() }

        elements.swapAt(0, elements.count - 1)
        let root = elements.removeLast()
        siftDown(from: 0)
        return root
    }

    private mutating func siftUp(from start: Int) {
        var child = start
        while child > 0 {
            let parent = (child - 1) / 2
            guard hasHigherPriority(elements[child], elements[parent]) else { return }
            elements.swapAt(child, parent)
            child = parent
        }
    }

    private mutating func siftDown(from start: Int) {
        var parent = start
        while true {
            let left = parent * 2 + 1
            guard left < elements.count else { return }
            let right = left + 1
            var candidate = left
            if right < elements.count && hasHigherPriority(elements[right], elements[left]) {
                candidate = right
            }
            guard hasHigherPriority(elements[candidate], elements[parent]) else { return }
            elements.swapAt(parent, candidate)
            parent = candidate
        }
    }
}

final class Solution {
    func findCheapestPrice(_ n: Int, _ flights: [[Int]], _ src: Int, _ dst: Int, _ k: Int) -> Int {
        // TODO: Implement
        fatalError("TODO: Implement")
    }
}

func runTests() {
    // TEST_VECTORS_BEGIN sha256:b385459593ac4e4946915f96f5d727c2afd40e58deb78294d28cef9dffedb5e6
    expectEqual(Solution().findCheapestPrice(4, [[0, 1, 100], [1, 2, 100], [2, 3, 100], [0, 2, 500]], 0, 3, 1), 600, "one-stop-limit")
    expectEqual(Solution().findCheapestPrice(4, [[0, 1, 100], [1, 2, 100], [2, 3, 100], [0, 2, 500]], 0, 3, 2), 300, "two-stop-limit")
    expectEqual(Solution().findCheapestPrice(3, [[0, 1, 100], [1, 2, 100], [0, 2, 500]], 0, 2, 0), 500, "direct-only")
    expectEqual(Solution().findCheapestPrice(3, [[0, 1, 100]], 0, 2, 1), -1, "unreachable")
    expectEqual(Solution().findCheapestPrice(1, [], 0, 0, 0), 0, "same-city")
    // EXCLUDED_VECTOR zero-cities: [0,[],0,0,0] | At least one city is provided.
    // EXCLUDED_VECTOR negative-stops: [2,[[0,1,5]],0,1,-1] | The stop limit is nonnegative.
    // TEST_VECTORS_END
    reportSuccess()
}

runTests()
